src/txn/txn.c

```diff
@@ -1253,6 +1258,26 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
     first_committed_upd_in_hs =
       first_committed_upd != NULL && F_ISSET(first_committed_upd, WT_UPDATE_HS);
 
+    /*
+     * We should not see the flags both set as we either successfully evict the page with prepared
+     * updates or not.
+     */
+    WT_ASSERT(session, !prepare_on_disk || !first_committed_upd);
+
+    /*
+     * If we see the first committed update has been moved to the history store, we must have done a
+     * successful reconciliation on the page but failed to evict it. Also reconciliation could not
+     * possibly empty the page because the prepared update is not globally visible. Therefore,
+     * reconciliation must have either split the page or done a page rewrite.
+     *
+     * In this case, we still need to resolve the prepared update as if we have successfully evicted
+     * the page because the value older than the prepared update has been written to the history
+     * store with the max timestamp.
+     */
+    WT_ASSERT(session,
+      !first_committed_upd_in_hs || page->modify->rec_result == WT_PM_REC_MULTIBLOCK ||
+        page->modify->rec_result == WT_PM_REC_REPLACE);
+
     /*
      * Marked the update older than the prepared update that is already in the history store to be
      * deleted from the history store.
```

Comment:
I find it difficult to follow the below logic.
At this stage if first_committed_upd_in_hs is true, then first_committed_upd is from history store, why we need to loop to locate the upd_followed_tombstone and check whether its txn_id != WT_TXN_ABORTED, as these updates are in history store or restored from history store, in both cases there will be no aborted update.

The logic below line 1301, trying to find the latest history store entry and trying to restore it, but that might not be required if we already have the first_committed_upd.

May be we can structure this code as switch: case to make it clear which cases is being handled based on the cases described in the above comment at line 1220?

Reply:
At this stage if first_committed_upd_in_hs is true, then first_committed_upd is from history store, why we need to loop to locate the upd_followed_tombstone and check whether its txn_id != WT_TXN_ABORTED, as these updates are in history store or restored from history store, in both cases there will be no aborted update.

No, in this case we are only handling the case that things are in history store but not restored from history store. We can have aborted updates in the middle as they are never successfully evicted.

The logic below line 1301, trying to find the latest history store entry and trying to restore it, but that might not be required if we already have the first_committed_upd.

We already check that by `!commit && prepare_on_disk`, we are not restoring from hs in case of `first_committed_upd_in_hs`

May be we can structure this code as switch: case to make it clear which cases is being handled based on the cases described in the above comment at line 1220?

It's a bit hard to do that. We will create some code duplication as a lot of the steps are shared.

Reply:
After a second thought, I think this is a good idea. Done as suggested.